src/os/windows/string_uniscribe.cpp

```diff
@@ -0,0 +1,610 @@
+	this->positions.resize(this->num_glyphs * 2 + 2);
+
+	int advance = 0;
+	for (int i = 0; i < this->num_glyphs; i++) {
+		this->positions[i * 2 + 0] = range.offsets[i].du + advance + x;
+		this->positions[i * 2 + 1] = range.offsets[i].dv;
+
+		advance += range.advances[i];
+	}
+	this->positions[this->num_glyphs * 2] = advance + x;
+}
+
+const int *UniscribeParagraphLayout::UniscribeVisualRun::GetGlyphToCharMap() const
+{
+	if (this->glyph_to_char == NULL) {
+		this->glyph_to_char = CallocT<int>(this->GetGlyphCount());
+
+		/* The char to glyph array contains the first glyph index of the cluster that is associated
+		 * with each character. It is possible for a cluster to be formed of several chars. */
+		for (int c = 0; c < (int)this->char_to_glyph.size(); c++) {
+			/* If multiple chars map to one glyph, only refer back to the first character. */
+			if (this->glyph_to_char[this->char_to_glyph[c]] == 0) this->glyph_to_char[this->char_to_glyph[c]] = c + this->start_pos;
+		}
+
+		/* We only marked the first glyph of each cluster in the loop above. Fill the gaps. */
+		int last_char = this->glyph_to_char[0];
+		for (int g = 0; g < this->GetGlyphCount(); g++) {
+			if (this->glyph_to_char[g] != 0) last_char = this->glyph_to_char[g];
+			this->glyph_to_char[g] = last_char;
+		}
+	}
+
+	return this->glyph_to_char;
+}
+
+
+/* virtual */ void UniscribeStringIterator::SetString(const char *s)
+{
+	const char *string_base = s;
+
+	this->utf16_to_utf8.clear();
+	this->str_info.clear();
+	this->cur_pos = 0;
+
+	/* Uniscribe operates on UTF-16, thus we have to convert the input string.
+	 * To be able to return proper offsets, we have to create a mapping at the same time. */
+	std::vector<wchar_t> utf16_str;     ///< UTF-16 copy of the string.
+	while (*s != '\0') {
+		size_t idx = s - string_base;
+
+		WChar c = Utf8Consume(&s);
+		if (c < 0x10000) {
+			utf16_str.push_back((wchar_t)c);
+		} else {
+			/* Make a surrogate pair. */
+			utf16_str.push_back((wchar_t)(0xD800 + ((c - 0x10000) >> 10)));
+			utf16_str.push_back((wchar_t)(0xDC00 + ((c - 0x10000) & 0x3FF)));
+			this->utf16_to_utf8.push_back(idx);
+		}
+		this->utf16_to_utf8.push_back(idx);
+	}
+	this->utf16_to_utf8.push_back(s - string_base);
+
+	/* Query Uniscribe for word and cluster break information. */
+	this->str_info.resize(utf16_to_utf8.size());
+
+	if (utf16_str.size() > 0) {
+		/* Itemize string into language runs. */
+		std::vector<SCRIPT_ITEM> runs = UniscribeItemizeString(&utf16_str[0], (int32)utf16_str.size());
+
+		for (std::vector<SCRIPT_ITEM>::const_iterator run = runs.begin(); runs.size() > 0 && run != runs.end() - 1; run++) {
+			/* Get information on valid word and character break.s */
+			int len = (run + 1)->iCharPos - run->iCharPos;
+			std::vector<SCRIPT_LOGATTR> attr(len);
+			ScriptBreak(&utf16_str[run->iCharPos], len, &run->a, &attr[0]);
+
+			/* Extract the information we're interested in. */
+			for (int c = 0; c < attr.size(); c++) {
```

Comment:
'size_t' is better than 'int' there (yeah gcc complained about signed/unsigned comparison